### ISSUE TITLE
Api suggestions?

### ISSUE THREAD

**Issue by DanielJoyce** [google/googletest-rust#3]
Is it every the case when we have a `verify_that!()` we don't want a test to eventually fail if that one fails? Why is there a need for `and_log_failure`?

I was expecting the api to be something more like:

```rust
#[google-test]
pub fn some_test() -> Result<...>{
verify_that!(12, eq(12)); // Okay
verify_that!(12, eq(13)); // Marked test as failure, but test continues
// This one passes and provides the return value which is determined by if any of the previous checks failed. 
verify_that!(10, eq(10))
}
```

Maybe it can be an alternate syntax instated of verify_that, or maybe the case of `verify_that!(...).and_log_error` could be shortened to `verify_and_continue!(...)`

**Comment by hovinen** [comment]
`verify_that!` just evaluates to a `Result`. It's up to the invoking the macro to decide what to do with that. If the macro would already take some action like outputting the failure message or marking the test failed, it could create surprising behaviour when the test is aborted immediately via the `?` operator.

I've considered adding `expect_that!` which would be analogous to GoogleTest's `EXPECT_*` macros. This would do what you describe by expanding to `verify_that!(...).and_log_failure()`. I'll bring this up with my colleagues.

**Comment by DanielJoyce** [comment]
Yes please! In the end, I can add that as a macro_rules! thing, but I think it would be handy.

**Comment by DanielJoyce** [comment]
I guess I don't understand the purpose of verify_that!() verifying something but then marking the test as a failure.

Reading on the google-test docs, they use assert to immediately abort tests ( because some core invariant has failed which means all subsequent tests make no sense ).

So I think it's fine and reasonable that ? aborts a test early, and otherwise it continues on to the end if possible, it would match the behavior 

Or, perhaps use the google_test distinction where ASSERT_ aborts the test immediately, and EXPECT_ allows the test to continue, but the test may fail at the end.

**Comment by DanielJoyce** [comment]
```
#[test]
the_get_cars_queries_returns_a_list_of_cars() {
    let cars = get_cars("hyundai", 2002)
   assert_eq!(200, cars.response.code) // Abort, no further checks make sense...
   ... extract the cars
   expect!( check that all the returned cars are hyundai )
   expect!( check they all model year 2002 )
}
```

I guess my point is that unless verify_that is used at the end, or with ?, then it apparently does nothing? It won't fail the test.

**Comment by hovinen** [comment]
That's right -- `verify_that!` acts as a pure function (provided the supplied matcher does too) which just evaluates to a `Result`. If one doesn't do anything with the `Result`, then nothing happens and the assertion result is discarded.

I know this feels unusual, since most test libraries just panic when the assertion fails. We structured it this way in part so that the caller has full control over what happens without being surprised by any side-effects. This can be useful in some more complex test setups with a lot of machinery between the test and the assertion (e.g., if the assertion is in a callback which is invoked somewhere).

I've discussed the matter with my colleagues and we're happy to introduce `expect_that!` and friends which would expand to `verify_that!(...).and_log_failure()` so as to reduce the boilerplate for non-fatal assertions. I plan to do so in time for the next release.
